### PR TITLE
Created babel-plugin-react-native-platform

### PR DESCRIPTION
**Pull request by EvanBacon**
# Why

Transform projects for the platform they're built to run on.

The following examples show what the output of a lexical block looks like after it's been transpiled with `expo build:web` (`babel-plugin-react-native-platform` & Webpack's Terser).

A current blocker is that right now it seems Metro just compiles code once for a single platform. This means the plugin breaks native.

### Input

```js
import { Platform } from '@unimodules/core';
let value;
if (Platform.OS === 'ios') {
  value = 'ios';
} else if (Platform.OS == 'android') {
  value = 'android';
} else if (Platform.OS === 'web') {
  value = 'web';
} else {
  value = 'default';
}
console.log('Output', value);
```

### Before

```js
var value,
  Platform = {
    OS: 'web',
    select(e) {
      return 'web' in e ? e.web : e.default;
    },
  };
(value =
  Platform.OS === 'ios'
    ? 'ios'
    : Platform.OS == 'android'
    ? 'android'
    : Platform.OS === 'web'
    ? 'web'
    : 'default'),
  console.log('Output', value);
```

### After

```js
console.log('Output', 'web');
```

### Input

```js
import { Platform } from 'react-native';
const value = Platform.select({
  ios: 'ios',
  android: 'android',
  web: 'web',
  default: 'default',
});
console.log('Output', value);
```

### Before

```js
var value = {
  OS: 'web',
  select(e) {
    return 'web' in e ? e.web : e.default;
  },
}.select({ ios: 'ios', android: 'android', web: 'web', default: 'default' });
console.log('Output', value);
```

### After

```js
console.log('Output', 'web');
```

### Input

```js
import { Platform } from 'react-native';
switch (Platform.OS) {
  case 'ios':
    console.log('iOS');
    break;
  case 'android':
    console.log('Android');
    break;
  case 'web':
    console.log('web');
    break;
  default:
    console.log('default');
    break;
}
```

### Before

```js
switch (
  {
    OS: 'web',
    select(e) {
      return 'web' in e ? e.web : e.default;
    },
  }.OS
) {
  case 'ios':
    console.log('iOS');
    break;
  case 'android':
    console.log('Android');
    break;
  case 'web':
    console.log('web');
    break;
  default:
    console.log('default');
}
```

### After

```js
console.log('web');
```

# How

Created `babel-plugin-universal-platforms`

# Test Plan

CI should pass

### CODE REVIEW

File: packages/babel-plugin-universal-platforms/src/__tests__/index-test.ts [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+      mode: 'development',
+    }
+  );
+
+  expect(code).toMatch('__DEV__');
+  expect(code).not.toMatch('process.env.NODE_ENV');
+});
+
+// react-native-web redefines this value.
+// Terser should shake it during the bundling.
```
Comment (ide) [inline packages/babel-plugin-universal-platforms/src/__tests__/index-test.ts]:
```suggestion
// Terser should remove it during the bundling.
```

File: packages/babel-plugin-universal-platforms/src/__tests__/index-test.ts [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+    }
+  );
+
+  expect(code).toMatch('DEV');
+  expect(code).not.toMatch('PROD');
+  expect(code).toMatchSnapshot();
+});
+
+// react-native-web redefines this value.
+// Terser should shake it during the bundling.
```
Comment (ide) [inline packages/babel-plugin-universal-platforms/src/__tests__/index-test.ts]:
```suggestion
// Terser should remove it during the bundling.
```

File: packages/babel-plugin-universal-platforms/src/__tests__/index-test.ts [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+  `,
+    {
+      platform: `web`,
+      mode: 'development',
+    }
+  );
+  expect(code).toMatch(`process.env.NODE_ENV = 'production'`);
+});
+
+it(`switch statement is converted for Terser`, () => {
```
Comment (ide) [inline packages/babel-plugin-universal-platforms/src/__tests__/index-test.ts]:
```suggestion
it(`converts switch-statement predicates for Terser`, () => {
```

File: packages/babel-plugin-universal-platforms/src/__tests__/index-test.ts [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+  `,
+    {
+      platform: `web`,
+      mode: 'development',
+    }
+  );
+  expect(code).toMatch('switch ("web")');
+});
+
+describe('if Statements', () => {
```
Comment (ide) [inline packages/babel-plugin-universal-platforms/src/__tests__/index-test.ts]:
```suggestion
describe('if statements', () => {
```

File: packages/babel-plugin-universal-platforms/src/__tests__/index-test.ts [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+      }
+    );
+
+    expect(code).not.toMatch('Platform.OS');
+    expect(code).toMatchSnapshot();
+  });
+});
+
+describe('Platform.select', () => {
+  it('should replace Platform.select with value for default case if there is no matching platform', () => {
```
Comment (ide) [inline packages/babel-plugin-universal-platforms/src/__tests__/index-test.ts]:
```suggestion
  it(`replaces Platform.select with the default value if no platform matches`, () => {
```

File: packages/babel-plugin-universal-platforms/src/__tests__/index-test.ts [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+        mode: 'development',
+      }
+    );
+
+    expect(code).toMatch(/const value = {\s*test: true\s*}/);
+    expect(code).not.toMatch('Platform.select');
+    expect(code).toMatchSnapshot();
+  });
+
+  it('should replace Platform.select with value for matching platform', () => {
```
Comment (ide) [inline packages/babel-plugin-universal-platforms/src/__tests__/index-test.ts]:
```suggestion
  it(`replaces Platform.select with the value for the matching platform`, () => {
```

File: packages/babel-plugin-universal-platforms/src/__tests__/index-test.ts [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+        mode: 'development',
+      }
+    );
+
+    expect(code).toMatch('const value = () => {};');
+    expect(code).not.toMatch('Platform.select');
+    expect(code).toMatchSnapshot();
+  });
+
+  it('should remove non-matching platforms but leave Platform.select', () => {
```
Comment (ide) [inline packages/babel-plugin-universal-platforms/src/__tests__/index-test.ts]:
```suggestion
  it(`removes non-matching platforms but leaves Platform.select if other platforms may match at runtime`, () => {
```

File: packages/babel-plugin-universal-platforms/src/__tests__/index-test.ts [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+      }
+    );
+
+    expect(code).toMatch('...defaults');
+    expect(code).toMatch('Platform.select');
+    expect(code).not.toMatch('ios');
+    expect(code).toMatchSnapshot();
+  });
+
+  it('should replace Platform.select with undefined if no cases are matching', () => {
```
Comment (ide) [inline packages/babel-plugin-universal-platforms/src/__tests__/index-test.ts]:
```suggestion
  it('replaces Platform.select with "undefined" if no platforms match`, () => {
```

File: packages/babel-plugin-universal-platforms/src/__tests__/index-test.ts [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+      {
+        platform: 'ios',
+        mode: 'development',
+      }
+    );
+
+    expect(code).toEqual('const value = undefined;');
+  });
+
+  it('should remove non-matching platforms but leave Platform.select with unknown cases 1', () => {
```
Comment (ide) [inline packages/babel-plugin-universal-platforms/src/__tests__/index-test.ts]:
```suggestion
  it(`removes non-matching platforms but leaves Platform.select with indeterminate platforms`, () => {
```

File: packages/babel-plugin-universal-platforms/src/__tests__/index-test.ts [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+      }
+    );
+
+    expect(code).toMatch("['andr' + 'oid']: () => {}");
+    expect(code).toMatch('Platform.select');
+    expect(code).not.toMatch('ios');
+    expect(code).toMatchSnapshot();
+  });
+
+  it('should remove non-matching platforms but leave Platform.select with unknown cases and default case', () => {
```
Comment (ide) [inline packages/babel-plugin-universal-platforms/src/__tests__/index-test.ts]:
```suggestion
  it(`removes non-matching platforms but leaves Platform.select with indeterminate platforms and the default case`, () => {
```

File: packages/babel-plugin-universal-platforms/src/__tests__/index-test.ts [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+      { platform: 'web', mode: 'development' }
+    );
+
+    expect(code).toMatch("['w' + 'eb']() {}");
+    expect(code).toMatch('Platform.select');
+    expect(code).toMatch("default: 'default',");
+    expect(code).toMatchSnapshot();
+  });
+
+  it('should remove non-matching platforms but leave Platform.select with method', () => {
```
Comment (ide) [inline packages/babel-plugin-universal-platforms/src/__tests__/index-test.ts]: I think we can do better here:
```
const value = Platform.select({ android() {} });
```
could become
```
const value = function() {};
```

Comment (ide) [inline packages/babel-plugin-universal-platforms/src/__tests__/index-test.ts]:
```suggestion
  it(`leaves Platform.select if the matching platform's value is a method`, () => {
```

File: packages/babel-plugin-universal-platforms/src/index.ts [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+  '||': (a, b) => a || b,
+  '!==': (a, b) => a !== b,
+  '===': (a, b) => a === b,
+  '!=': (a, b) => a != b,
+  '==': (a, b) => a == b,
+};
+
+const isLiteral = (node: t.Node): boolean =>
+  t.isLiteral(node) || (t.isIdentifier(node) && node.name === 'undefined');
+// /Literal$/.test(node.type) || (node.type === 'Identifier' && node.name === 'undefined');
```
Comment (ide) [inline packages/babel-plugin-universal-platforms/src/index.ts]:
```suggestion
```

File: packages/babel-plugin-universal-platforms/src/index.ts [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+export type Options = {
+  platform: string;
+  mode: string;
+};
+
+export default function(api: any, options: Options) {
+  const { platform, mode } = options;
+  const isDevelopment = mode !== 'production';
+  if (!platform) {
+    throw new Error('babel-plugin-universal-platforms: platform option must be defined');
```
Comment (ide) [inline packages/babel-plugin-universal-platforms/src/index.ts]:
```suggestion
    throw new Error('babel-plugin-universal-platforms: "platform" option must be defined');
```

File: packages/babel-plugin-universal-platforms/src/index.ts [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+  }
+
+  const collapseTestVisitor = {
+    /**
+     * Transforms static ID values for Terser to shake
```
Comment (ide) [inline packages/babel-plugin-universal-platforms/src/index.ts]:
```suggestion
     * Transforms static ID values for Terser to analyze
```

File: packages/babel-plugin-universal-platforms/src/index.ts [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+      if (!t.isVariableDeclarator(p.parent)) {
+        if (p.node.name === '__PLATFORM__') {
+          p.replaceWith(t.stringLiteral(platform));
+        } else if (p.node.name === '__DEV__') {
+          p.replaceWith(t.booleanLiteral(isDevelopment));
+        }
+      }
+    },
+    /**
+     * Transforms member expressions for Terser to shake
```
Comment (ide) [inline packages/babel-plugin-universal-platforms/src/index.ts]:
```suggestion
     * Transforms member expressions for Terser to analyze
```

File: packages/babel-plugin-universal-platforms/src/index.ts [hunk 16]
```diff
@@ -0,0 +1,184 @@
...
+        if (p.matchesPattern('Platform.OS')) {
+          p.replaceWith(t.stringLiteral(platform));
+        } else if (p.matchesPattern('process.env.NODE_ENV')) {
+          p.replaceWith(t.stringLiteral(mode));
+        }
+      }
+    },
+    UnaryExpression: {
+      /**
+       * Transforms redundant boolean expressions
```
Comment (ide) [inline packages/babel-plugin-universal-platforms/src/index.ts]: do we need to do this? shouldn't terser take care of this?

File: packages/babel-plugin-universal-platforms/src/index.ts [hunk 17]
```diff
@@ -0,0 +1,184 @@
...
+              defaultCase = property;
+            } else if (property.key.name === platform) {
+              targetCase = property;
+            }
+          } else {
+            canStripPlatformSelect = false;
+            additionalProperties.push(property);
+          }
+        });
+        // If we got exact mach, we can strip the rest
```
Comment (ide) [inline packages/babel-plugin-universal-platforms/src/index.ts]:
```suggestion
        // If we got an exact match, we can strip the rest
```

File: packages/babel-plugin-universal-platforms/src/index.ts [hunk 18]
```diff
@@ -0,0 +1,184 @@
...
+       */
+
+      exit(p: NodePath<t.UnaryExpression>) {
+        if (p.node.operator === '!' && isLiteral(p.node.argument)) {
+          const literal = p.node.argument as ValueLiteral;
+          p.replaceWith(t.booleanLiteral(!literal.value));
+        }
+      },
+    },
+    'BinaryExpression|LogicalExpression': {
```
Comment (ide) [inline packages/babel-plugin-universal-platforms/src/index.ts]: shouldn't terser take care of this as well?

**Review by ide — CHANGES_REQUESTED** [review]

**Comment by EvanBacon** [comment]
Closing for now

**Pull request closed without merging**